File: src/Utils/mock_org_chart_data.ts
// Import Utils
import { generate_id, range } from "./generate_id";
import { DoublyLinkedList } from "../TidyTree/DoublyLinkedList";

export interface MockCard {
  id: number;
  parent: number;
  children: string[];
  width: number;
  height: number;
  collapse: boolean;
}

function build_card(
  width: number | Array<number>,
  height: number | Array<number>,
  parent: number = -1,
  randomCollapse: boolean,
): MockCard {
  let _width = width instanceof Array ? range(width[0], width[1]) : width;
  let _height = height instanceof Array ? range(height[0], height[1]) : height;
  let id = generate_id();
  const collapse = randomCollapse ? false : Math.random() < 0.5;
  return {
    id: id,
    children: [],
    width: _width,
    height: _height,
    parent,
    collapse,
  };
}

export function mock_org_chart_data(
  count: number = 1,
  max_child?: number,
  is_range = false,
  width: number | Array<number> = 200,
  height: number | Array<number> = 100,
  randomCollapse = false,
): MockCard[] {
  max_child = max_child || Math.sqrt(count);
  let result = [];
  let queue = new DoublyLinkedList<any>();

  // generated leaf count
  let remain_count = count - 1;

  // build the root leaf
  let root = build_card(width, height, -1, randomCollapse);

  result.push(root);
  queue.push(root);

  while (!queue.is_empty()) {
    let node = queue.shift();
    let children: number[] = [];
    let children_count = Math.min(max_child, remain_count);
    if (is_range) {
      children_count = range(0, children_count);
    }

    for (let i = 0; i < children_count; i++) {
      let card = build_card(width, height, node.id, randomCollapse);
      children.push(card.id);
      queue.push(card);
      result.push(card);
      remain_count--;
    }

    node!.children = children;

    if (remain_count <= 0) {
      return result;
    }
  }

  return result;
}

File: rust/src/utils/mock_org_chart_data.rs
// use rust std
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;

// use third party lib
use rand::Rng;

//  use local types
use crate::utils::generate_id::GenerateID;

#[derive(Debug, Clone)]
pub struct MockChartData {
    pub id: i64,
    pub width: f32,
    pub height: f32,
    pub children: Vec<i64>,
}

pub fn build_card(generate_id: &mut GenerateID, width: f32, height: f32) -> MockChartData {
    let id = generate_id.get_next_id();

    MockChartData { id, width, height, children: vec![] }
}

pub fn range(min: i64, max: i64) -> i64 {
    let mut rng = rand::thread_rng();
    rng.gen_range(min..max)
}

pub fn mock_org_chart_data(count: i64, max_child: i64, is_range: bool, width: f32, height: f32) -> Vec<MockChartData> {
    let mut result = vec![];
    let mut queue = VecDeque::new();
    let mut generate_id = GenerateID::new();

    // generated leaf count
    let mut remain_count = count - 1;

    // build the root leaf
    let root = Rc::new(RefCell::new(build_card(&mut generate_id, width, height)));

    result.push(Rc::clone(&root));
    queue.push_back(Rc::clone(&root));

    while !queue.is_empty() {
        let node = queue.pop_front().unwrap();

        let mut children = vec![];
        let mut child_count = max_child.min(remain_count);

        if is_range {
            child_count = range(0, child_count)
        }

        for _ in 0..child_count {
            remain_count -= 1;

            let card = Rc::new(RefCell::new(build_card(&mut generate_id, width, height)));
            children.push(card.borrow().id);
            queue.push_back(Rc::clone(&card));
            result.push(Rc::clone(&card));
        }

        node.borrow_mut().children = children;

        if remain_count <= 0 {
            return result.iter().map(|x| x.borrow().clone()).collect();
        }
    }

    result.iter().map(|x| x.borrow().clone()).collect()
}
